packages/engine/Source/Shaders/AtmosphereCommon.glsl

```diff
@@ -50,14 +48,46 @@ void computeScattering(
         return;
     }
 
+    // To deal with smaller values of PRIMARY_STEPS (e.g. 4)
+    // we implement a split strategy: sky or horizon.
+    //
+    // For performance reasons, instead of a if/else branch
+    // a soft choice is implemented through a weight 0.0 <= w_stop_gt_lprl <= 1.0
+    float lprl = length(primaryRayLength);
+    float x = 1e-7 * primaryRayAtmosphereIntersect.stop / lprl;
+    // w_stop_gt_lprl: similar to (1+tanh)/2
+    // Value close to 0.0: close to the horizon
+    // Value close to 1.0: above in the sky
+    float w_stop_gt_lprl = max(0.0,min(1.0,(1.0 + x * ( 27.0 + x * x ) / ( 27.0 + 9.0 * x * x ))/2.0));
+
     // The ray should start from the first intersection with the outer atmopshere, or from the camera position, if it is inside the atmosphere.
+    float start_0 = primaryRayAtmosphereIntersect.start;
     primaryRayAtmosphereIntersect.start = max(primaryRayAtmosphereIntersect.start, 0.0);
     // The ray should end at the exit from the atmosphere or at the distance to the vertex, whichever is smaller.
-    primaryRayAtmosphereIntersect.stop = min(primaryRayAtmosphereIntersect.stop, length(primaryRayLength));
+    primaryRayAtmosphereIntersect.stop = min(primaryRayAtmosphereIntersect.stop, lprl);
+
+
+    // Distinguish inside or outside atmosphere (outer space)
+    // Reasons: 
+    // (1) from outer space we have to use the original implementation to get an unmodified/acceptable rendering
+    // (2) within atmosphere we need a speedup
+    float x_o_a = start_0 - ATMOSPHERE_THICKNESS; // ATMOSPHERE_THICKNESS used as an ad-hoc constant, no precise meaning here, only the order of magnitude matters
+    float w_inside_atmosphere = 1.0 - max(0.0,min(1.0,(1.0 + x_o_a * ( 27.0 + x_o_a * x_o_a ) / ( 27.0 + 9.0 * x_o_a * x_o_a ))/2.0)); // similar to 1.0 - (1+tanh)/2
+    
+    // (1) Outside of the atmosphere: original (16.0,4.0) values for good rendering
+    // (2) Inside atmosphere: smaller (4.0,2.0) values for a speedup
+    int PRIMARY_STEPS = int(16.0 - w_inside_atmosphere * 12.0); // Number of times the ray from the camera to the world position (primary ray) is sampled.
+    int LIGHT_STEPS   = int(4.0 - w_inside_atmosphere * 2.0); // Number of times the light is sampled from the light source's intersection with the atmosphere to a sample position on the primary ray.
 
     // Setup for sampling positions along the ray - starting from the intersection with the outer ring of the atmosphere.
-    float rayStepLength = (primaryRayAtmosphereIntersect.stop - primaryRayAtmosphereIntersect.start) / float(PRIMARY_STEPS);
     float rayPositionLength = primaryRayAtmosphereIntersect.start;
+    // (1) Outside of the atmosphere: original implementation
```

Comment:
Once this PR is merged, someone reading the code will not know what the "original implementation" was. These comments could simply state what is different in each case. For example:

```glsl
// (1) Outside the atmosphere: constant rayStepLength
```